Add the ability to handle templates

d3adb5 opened the pull request:
Add the ability to handle templated bookmarks by creating a ZLE widget to jump
to the next template field.

Commands containing parts surrounded by double curly braces (`{{`) are not
considered "templated bookmarks", and smarky now has a ZLE widget to jump to the
next field with Ctrl + G.

Might suffer extensive modification in the future to make it more friendly,
highlighting the template fields so they're clearly visible to the user instead
of just replacing them entirely upon pressing the "next-field" keybind.

Closes #3.

d3adb5 committed:
fix: don't change cmdline if no mark is selected

Avoid changing the command line if no bookmark is selected by the user
through FZF.

d3adb5 committed:
feat: jump to next tmpl field with C-g

Add a ZLE widget that allows one to jump to the next "template" field
with the keybind Ctrl + G. Fields are identified by text enclosed in
double curly braces ({{ like this }}), and do not have any special
syntax inside of them.

d3adb5 committed:
docs: add information on templates in README

Add brief information on templated bookmarks to the repository's
README.md.

d3adb5 committed:
docs: add gifs showcasing smarky's behavior

Add GIFs to the README showcasing smarky's behavior, including its ZLE
widgets.

d3adb5 committed:
docs: add unlicense

Add the LICENSE notice showing this software is part of public domain.

viniciustrainotti commented on README.md:
```diff
@@ -22,37 +22,47 @@ smarky depends on: `fzf`, `sqlite3`, `bat` (optional)
+
+![smarky templated commands](https://user-images.githubusercontent.com/8827351/174421186-028d50ac-8df2-440a-8c4d-a6ccb8a1f4a9.gif)
+
+You can even take this opportunity to invoke more well-behaved widgets.
```
I don't understand this final phrase... Another widgets out smarky or do you refer two widgests from smarky?

d3adb5 replied:
I guess it did end up being a little vague. I'll adjust the phrasing in a new commit. To clarify: what I meant by that is that since the moment you retrieve something from smarky by using the `fzf-smarky-pick-command` widget control is forfeited by smarky and passed back over to ZSH, you can invoke as many widgets as you like.

This GIF hopefully makes more sense than my words:

![smarky-widgets-thing](https://user-images.githubusercontent.com/8827351/174508861-ddb2f336-b511-4e34-9030-c91e7a874203.gif)

d3adb5 replied:
I changed the phrasing in [3bf50a1](https://github.com/d3adb5/smarky/pull/4/commits/3bf50a1c2723e8a7412b9db78959aed9335ff9cb). What do you think? Not too happy with it, so I'm very open to suggestions.

viniciustrainotti replied:
My suggestion is remove the latest phrase and kept the GIF latest "thing/way" ... but this commit 3bf50a1 I bring it's ok right now ... See what looks better and give a asked here

d3adb5 replied:
I think I get what you mean. The sentence might be a little confusing, so it's best to just have the GIF and let people try using tools they know for themselves. I removed the paragraph I added in 06d20ae.

viniciustrainotti commented on zsh/widgets.zsh:
```diff
@@ -8,9 +8,20 @@ fzf-smarky-pick-command() {
+  setopt localoptions noglobsubst noposixbuiltins pipefail 2>/dev/null
+  [ -z "$RBUFFER" ] && CURSOR=0
+  local subtracted="${RBUFFER#*\{\{}" offset
+  offset=$(( ${#RBUFFER} - ${#subtracted} - 2 + $CURSOR ))
```
What's number 2 in calculation offset? Display? For understanding declare what 2 is, for example:

```sh
local PI=3.14
offset=$(( ${#RBUFFER} - ${#subtracted} - $PI + $CURSOR ))
```

d3adb5 replied:
In this particular case the `2` is the length of the search string (`{{`) being used for templates.

Since the search string itself was hardcoded (see `${RBUFFER#*\{\{}`), I thought hardcoding the number would've been alright in the end. Adding a comment for this feels unnecessary, but so does creating a string for it. The alternative looks particularly verbose to me:

```sh
local searchstr='{{' subtracted="${RBUFFER#*$searchstr}" offset
offset=$(( ${#RBUFFER} - ${#subtracted} - ${#searchstr} + $CURSOR ))
```

Especially since it might be replicated for the closing `}}`.

What do you think? Should I make a variable for this anyway?

viniciustrainotti replied:
I don't think it's unnecessary, if you're going to leave it as a public domain, every reference, variable to the code is valid. I believe that this second code is more readable than the first, but this is my simple opinion, go what you think is best for you. 👍🏼

d3adb5 replied:
The example I gave was wrong in the end, since you can't really use variables in the command you're using to define them. If you try that, you'll end up with `subtracted` having the same value as `RBUFFER`, since it can't know the contents of `$searchstr` until the command itself is finished.

That said, what I arrived at doesn't look too verbose. You're right, it does seem preferable to have variables for it now that I take a look at it. Here's what I got in the end:

```sh
setopt localoptions noglobsubst noposixbuiltins pipefail 2>/dev/null
[ -z "$RBUFFER" ] && CURSOR=0
local fieldstart='{{' fieldend='}}'
local subtracted="${RBUFFER#*$fieldstart}" offset
offset=$(( ${#RBUFFER} - ${#subtracted} - ${#fieldstart} + $CURSOR ))
CURSOR=offset; RBUFFER="${RBUFFER#*$fieldend}"
```

I don't know if it being public domain bears a lot of weight in this, but I agree with you that it is more readable this way. The change is in f1522ba.

**EDIT:** There was a leftover line I used to figure out what was happening. I removed it in ef09bb1.

d3adb5 committed:
docs: clarify reference to other widgets

Clarify the reference to invoking other widgets while editing the
command retrieved from smarky.

d3adb5 committed:
style: use variables for field delimiters

Use variables to store the delimiters used to identify template fields.

d3adb5 committed:
docs: remove confusing mention of other widgets

Remove vaguely worded, potentially confusing mention of other ZLE
widgets that could end up making the README harder to follow. People who
use ZLE widgets are probably aware of the possibility already.

d3adb5 committed:
fix: remove leftover 'debugging' echo

Remove leftover echo line used for debugging to figure out issues with
variable expansion. Ouch, how did I not catch that?

viniciustrainotti reviewed (approved):
🎉
